src/components/action-popover/action-popover-item/action-popover-item.component.tsx

```diff
@@ -139,19 +143,106 @@ export const ActionPopoverItem = ({
   const mouseEnterTimer = useRef<NodeJS.Timeout | null>(null);
   const mouseLeaveTimer = useRef<NodeJS.Timeout | null>(null);
 
+  const currentSubmenuAlignment = () => {
+    if (submenuPosition === "left") {
+      if (leftAlignmentState) {
+        return "right";
+      }
+      return "left";
+    }
+    return "right";
+  };
+
+  function getContainerPosition(
+    itemRef: React.RefObject<HTMLElement>,
+    containerSubmenuRef: React.RefObject<HTMLElement>,
+    containerPlacement: "bottom" | "top"
+  ): ContainerPosition | undefined {
+    /* istanbul ignore if */
+    if (!itemRef.current || !containerSubmenuRef.current) return undefined;
+
+    const { offsetWidth: submenuWidth } = containerSubmenuRef.current;
+
+    const horizontalAlignmentValue = currentSubmenuAlignment() === "left";
+    const leftValue = horizontalAlignmentValue ? -submenuWidth : "auto";
+    const rightValue = horizontalAlignmentValue ? "auto" : -submenuWidth;
+    const yPositionName = containerPlacement === "top" ? "bottom" : "top";
+
+    return {
+      left: leftValue,
+      [yPositionName]: "calc(-1 * var(--spacing100))",
+      right: rightValue,
+    };
+  }
+
+  const retrieveVariantType = () => {
+    if (!iconState && !caretState && !icon && !submenu) {
+      return 1;
+    }
+    if (!iconState && caretState && !icon && !submenu) {
+      return 2;
+    }
+    if (!iconState && caretState && !icon && submenu) {
+      return 3;
+    }
+    if (iconState && !caretState && !icon && !submenu) {
+      return 4;
+    }
+    if (iconState && !caretState && icon && !submenu) {
+      return 5;
+    }
+    if (iconState && caretState && icon && submenu) {
+      return 6;
+    }
+    if (iconState && caretState && icon && !submenu) {
+      return 7;
+    }
+    if (iconState && caretState && !icon && submenu) {
+      return 8;
+    }
+    if (iconState && caretState && !icon && !submenu) {
+      return 9;
+    }
+    return 1;
+  };
+
   useEffect(() => {
     if (!isOpenPopover) {
       setOpen(false);
     }
   }, [isOpenPopover]);
 
+  useEffect(() => {
+    if (icon) {
+      setIconState?.(true);
+    }
+    if (submenu) {
+      setCaretState?.(true);
+    }
+  }, [icon, setCaretState, setIconState, submenu]);
+
+  useEffect(() => {
```

Comment:
**suggestion**: exactly the same comment here regarding `isLeftAligned`/`leftAlignmentState` as with `iconState` and `caretState` in the comment above - we can surely avoid any of this state and passing it down as props, when all we need to do is use the `isLeftAligned` state which is internal to `ActionPopoverItem`.

Reply:
A similar approach was used with `isLeftAligned` and `leftAlignmentState` as `isLeftAligned` is only false when a childs `submenu` width is smaller than space available on the left-hand side of the container. Technically, `isLeftAligned` will never be false for children without a `submenu`. Therefore, without this approach of changing state once to true for all siblings, only some children would switch to the right. With this approach, as soon as the criteria is met for one child, every single child switches to the right. Which is a lot cleaner imo 👍 , happy to discuss this approach, and the other sibling state approaches too though.

Reply:
question: should we handle for the potential that icons could be added/ removed dynamically as well, this current approach doesn't cater for that? Happy to leave it as is for now but figured I'd ask and see what you and Robin thought